### PR TITLE
Using the new core type system

### PR DESCRIPTION
**Pull request by magnatelee**
This PR updates the cuNumeric code to use the new core type system in https://github.com/nv-legate/legate.core/pull/697. Though this PR updates almost every C++ file in the repository, the majority of changes is just to catch up the changes in type codes (`legate::LegateTypeCode` => `legate::Type::Code`).

### CODE REVIEW

**Review by jjwilke — APPROVED** [review]
LGTM. A few minor nits, but no blockers.

File: cunumeric/runtime.py [hunk 1]
```diff
@@ -131,19 +121,14 @@ def _unload_cudalibs(self) -> None:
...
-            code = type_system[value_dtype].code
-            dtype = type_system.add_type(arg_dtype, arg_dtype.itemsize, code)
-
-            for redop in CuNumericRedopCode:
-                redop_id = self.legate_context.get_reduction_op_id(
-                    redop.value * legion.MAX_TYPE_NUMBER + code
-                )
-                dtype.register_reduction_op(redop, redop_id)
+    def get_arg_dtype(self, value_dtype: ty.Dtype) -> ty.Dtype:
+        if value_dtype in self._arg_dtypes:
```
[the remark this answers is not in the document]
Comment (bryevdv) [reply cunumeric/runtime.py:125]: Is that true? The implementations look nearly identical and both call `_Py_dict_lookup` 

https://github.com/python/cpython/blob/main/Objects/dictobject.c#L3224-L3276

If anything `dict_get_impl` has slighly more work do to a `Py_NewRef(val)` at the end.

[the remark this answers is not in the document]
Comment (bryevdv) [reply cunumeric/runtime.py:125]: Oh, GH omitted the full context, I thought the suggestion was in reference to the single line 
```
if value_dtype in self._arg_dtypes:
```
I usually try to highlight blocks and use the `Add a suggestion` button to get a full diff in the comment, or put the comment on the last line all the code lines I want to talk about, for this reason.

File: cunumeric/deferred.py [hunk 2]
```diff
@@ -653,7 +646,7 @@ def _advanced_indexing_with_boolean_array(
                 DeferredArray,
                 self.runtime.create_empty_thunk(
                     out_shape,
-                    rhs.dtype,
+                    rhs.base.type,
```
[the remark this answers is not in the document]
Comment (magnatelee) [reply cunumeric/deferred.py:649]: Yeah I think that's the right thing to do, but unfortunately, `EagerArray` is being used for whatever the deferred implementation cannot handle, including non-primitive types. I think we can and should stop pretending that we support those types, but I'd rather want to do it in a follow-on PR and not this one. Let's stash it for now.

File: src/cunumeric/arg_redop_register.cu [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "cunumeric/arg_redop_register.h"
+
+extern "C" {
+
+void cunumeric_register_reduction_op(int32_t type_uid, int32_t _elem_type_code)
```
[the remark this answers is not in the document]
Comment (magnatelee) [reply src/cunumeric/arg_redop_register.cu:21]: We need device reduction operators for the dma engine to perform reductions using CUDA kernels. There's a macro in legate core that determines whether a device reduction operator should be registered: https://github.com/nv-legate/legate.core/blob/branch-23.05/src/core/runtime/context.inl#L39-L76 Unfortunately, I don't think we can compile the device operator using a normal C++ compiler nor can we have only the cc file.

File: cunumeric/runtime.py [hunk 4]
```diff
@@ -87,28 +80,25 @@ def __init__(self, legate_context: LegateContext) -> None:
...
+        self._arg_dtypes: dict[ty.Dtype, ty.Dtype] = dict()
 
-    def get_point_type(self, n: int) -> np.dtype[Any]:
-        type_system = self.legate_context.type_system
-        point_type = np.dtype(",".join(("i8",) * n))
-        if point_type not in type_system:
-            raise ValueError(f"there is no point type registered for {n}")
-        return point_type
+    def get_point_type(self, n: int) -> ty.Dtype:
+        if n in self._point_dtypes:
```
[the remark this answers is not in the document]
Comment (magnatelee) [reply cunumeric/runtime.py]: https://github.com/nv-legate/cunumeric/pull/903/commits/af77875526de6589de5784f802b7fce1690c9233

File: cunumeric/runtime.py [hunk 5]
```diff
@@ -87,28 +80,25 @@ def __init__(self, legate_context: LegateContext) -> None:
...
             self._load_cudalibs()
 
-    def _register_dtypes(self) -> None:
-        type_system = self.legate_context.type_system
-        for numpy_type, core_type in SUPPORTED_DTYPES.items():
-            type_system.make_alias(np.dtype(numpy_type), core_type)
-
-        for dtype in _CUNUMERIC_DTYPES:
-            type_system.add_type(dtype[0], dtype[1], dtype[2])
+        self._point_dtypes: dict[int, ty.Dtype] = dict()
```
[the remark this answers is not in the document]
Comment (magnatelee) [reply cunumeric/runtime.py]: https://github.com/nv-legate/cunumeric/pull/903/commits/af77875526de6589de5784f802b7fce1690c9233

File: cunumeric/runtime.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (magnatelee) [reply cunumeric/runtime.py]: https://github.com/nv-legate/cunumeric/pull/903/commits/af77875526de6589de5784f802b7fce1690c9233

File: cunumeric/deferred.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jjwilke) [reply cunumeric/deferred.py:649]: 👍

**Review by jjwilke — APPROVED** [review]
LGTM

File: src/cunumeric/sort/sort_cpu.inl [hunk 6]
```diff
@@ -552,7 +552,7 @@ void sample_sort_nd(SortPiece<legate_type_of<CODE>> local_sorted,
       /*comm::coll::collAllgather(p_samples + num_samples_l * my_sort_rank,
                                 p_samples,
                                 num_samples_l * sizeof(SegmentSample<VAL>),
-                                comm::coll::CollDataType::CollUint8,
+                                comm::coll::CollDataType::Code::CollUint8,
```
Comment (manopapad) [inline src/cunumeric/sort/sort_cpu.inl:555]: Overager search/replace

**Pull request merged**